[Infrix] iP

Infrix opened the pull request:
# Heard of Duke? Like Cats? Now you have both

> "This chatbot is amazing!" - my friends 

~~No longer just a boring task manager~~ Now a cute task manger, Lulu comes with the following characteristics!
- text-based
- easy to learn
- extremely fast to use
- cat UI

Here is the concise main method as a sneak peek
```java
public class Main {
    public static void main(String[] args) throws IndexOutOfBoundsException {
        Lulu lulu = new Lulu("./data/lulu.txt");
        lulu.run();
    }
}
```

It also has the following commands
```
LIST
MARK 
UNMARK 
DELETE 
DEADLINE 
EVENT
TODO 
BYE
FIND 
```

Want this task manager? Do the following steps!
1. Download it from [here!](https://github.com/Infrix/ip/releases)
2. Add a data folder if you would like a storage functionality
3. Double click it
3. Add your tasks
4. Let him manage your tasks for you :smile_cat:
 
Features:
- [x] Managing tasks
- [ ] GUI
- [ ] Becoming an actual cat

Commit:
Add Gradle support

Infrix committed:
Implemented skeletal versions of Lulu that starts by greeting the user, echos commands by user and exits when the user types bye.

Infrix committed:
Added the ability to store any text entered by the user and display them back to the user when "list" is entered.

Infrix committed:
Add the ability to mark tasks as done, add the ability to mark tasks as undone, add class Task that encapsulates information related to tasks

Infrix committed:
Add Todo, Event, Deadline, all inheriting from Task class

Infrix committed:
Add automated Text UI Testing

Infrix committed:
Add exceptions for Lulu to handle when dealing with incorrect inputs entered by the user

Infrix committed:
Add delete command for deleting tasks from list, implement ArrayList for tasks, handle IndexOutOfBounds exception

Infrix committed:
Add file read, file write, change logic of chatbot

Infrix committed:
Add date for deadline and event

Infrix committed:
Merge branch 'branch-Level-7'

Infrix committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Deadline.java
#	src/main/java/Event.java

Add bug fixes, ability to recognise date or string

Infrix committed:
Refactor code to utilise more OOP, add enums, add commands

Infrix committed:
Organise into packages, command, task, exception
Include private modifiers for attributes

Infrix committed:
Merge branch 'add-gradle-support' of https://github.com/Infrix/ip

Infrix committed:
Add JUnit tests to test the behavior of code

Infrix committed:
Package app as an executable JAR file

Infrix committed:
Add JavaDoc comments

Infrix committed:
Tweak switch to comply with coding standard

Infrix committed:
Add new find command for users to find a task by searching for a keyword

Infrix committed:
Merge branch 'branch-A-JavaDoc'

Infrix committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/lulu/TaskList.java
#	src/main/java/lulu/command/Parser.java

Infrix committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/lulu/TaskList.java
#	src/main/java/lulu/command/Parser.java

glozxi commented on src/main/java/lulu/TaskList.java:
```diff
@@ -0,0 +1,92 @@
+    }
+
+    public String getTaskDescription(int taskNumber) {
+        return this.list.get(taskNumber - 1).toString();
```
Perhaps use a separate method for getting the correct index to make it clear why `taskNumber - 1` needs to be done?

glozxi commented on src/main/java/lulu/task/Event.java:
```diff
@@ -0,0 +1,66 @@
+    public Event(String description, String from, String to) {
+        super(description);
+        try {
+            if (from.split("-").length == 3) {
```
Perhaps add a comment explaining why the specific numbers are used or use named constants?

glozxi commented on src/main/java/lulu/Lulu.java:
```diff
@@ -0,0 +1,54 @@
+            } catch (IndexOutOfBoundsException e) {
+                ui.showLine();
+                ui.showOutOfBounds();
+                ui.showLine();
```
Perhaps you could create a separate method to print a message between 2 lines?

glozxi commented on src/main/java/lulu/command/Parser.java:
```diff
@@ -0,0 +1,59 @@
+            String restOfCommand =
+                    i == -1
+                            ? ""
+                            : inputCommand.substring(i + 1, size);
```
Maybe add a comment to explain how the parsing works?

Infrix committed:
Basic GUI

Infrix committed:
Add GUI, add use of varargs

Infrix committed:
Merge branch 'branch-GUI'

pangrwa commented on src/main/java/lulu/Lulu.java:
```diff
@@ -0,0 +1,73 @@
+public class Lulu {
+    private static final String NAME = "./data/lulu.txt";
+    private TaskList tasks;
+    private UI ui;
```
Maybe perhaps consider changing your class name to  `Ui` instead? Since Ui is an abbreviation.
```suggestion
    private Ui ui;
```

pangrwa commented on src/main/java/lulu/Lulu.java:
```diff
@@ -0,0 +1,73 @@
+     * Replace this stub with your completed method.
+     */
+    public String getResponse(String fullCommand) {
+        if (!this.isSaveLoaded) {
```
Perhaps a more intuitive variable name here?

pangrwa commented on src/main/java/lulu/Storage.java:
```diff
@@ -0,0 +1,72 @@
+     *
+     * @param list the array of string to be written
+     */
+    public void writeSave(ArrayList<String> list) {
```
Perhaps you might want to consider to change `list` into `tasks` to show that its plural.

pangrwa commented on src/main/java/lulu/command/Command.java:
```diff
@@ -0,0 +1,24 @@
+    public abstract String execute(TaskList tasks, UI ui, Storage storage);
+
+    /**
+     * @return true if the command is bye, false otherwise.
```
Should there be a method explanation?

pangrwa commented on src/main/java/lulu/command/LoadCommand.java:
```diff
@@ -0,0 +1,22 @@
+import lulu.Storage;
+
+public class LoadCommand extends Command {
+    public LoadCommand() {
```
Since it's default for Java to use an empty constructor, should this be removed?

pangrwa reviewed (commented):
Overall, I found your code nice and concise! I like how you abstracted most of the classes out, hence it was easy for me to view your code. However, I noted some minor coding standard related issues such as some variables name being a little ambiguous. Keep it up!

OliviaJHL commented on src/main/java/lulu/Lulu.java:
```diff
@@ -0,0 +1,73 @@
+    private boolean isSaveLoaded = false;
+
+    public Lulu() {
+        this.ui = new UI();
```
maybe better to use Ui instead of UI

OliviaJHL commented on src/main/java/lulu/task/Deadline.java:
```diff
@@ -0,0 +1,48 @@
+
+import java.time.LocalDate;
+import java.time.format.DateTimeParseException;
+
```
Maybe can add JavaDoc in Deadline.java, Event.java, Task.java, and Todo.java

OliviaJHL commented on src/main/java/lulu/task/Deadline.java:
```diff
@@ -0,0 +1,48 @@
+    @Override
+    public String toMemory() {
+        int i = this.isDone ? 1 : 0;
+        String s;
```
maybe can give specific name instead of using i and s, such as status, str...only suggestion...

Infrix committed:
Add assertions

No assertions to account for programmer mistakes

Add assertions for TaskList

Other classes are more prone to user mistakes, TaskList is the most susceptible to
programmer mistakes

Infrix committed:
Improve code quality

Event and Deadline have magic numbers in constructor

Use a named constant to make the code easier to understand

Infrix committed:
Merge pull request #1 from Infrix/branch-A-Assertions

Add assertions

Infrix committed:
Merge pull request #2 from Infrix/branch-A-CodeQuality

Improve code quality

Infrix committed:
No way to update tasks

Add new command to update tasks

Infrix committed:
Merge branch 'branch-BCD-Extension'

Infrix committed:
Missing javadoc comments, missing user guide, user has to create a data directory for
save system to work

Add javadoc comments to all non self-explanatory methods and classes
Add userguide
Storage automatically creates a data directory
